File: app/backend/src/routes/services.ts
import {Router, Response} from 'express';
import {requireAuth} from '../middleware/auth';
import {AuthenticatedRequest} from '../types';
import {discoverServices} from '../services/kubernetes';
import {logger} from '../services/logger';

const router = Router();

router.get('/', requireAuth, async (_req: AuthenticatedRequest, res: Response) => {
    try {
        const services = await discoverServices();
        res.json(services);
    } catch (err) {
        logger.error('Service discovery failed', {error: err});
        res.status(500).json({error: 'Failed to discover services'});
    }
});

export default router;

File: app/backend/internal/handlers/services.go
package handlers

import (
	"net/http"

	"github.com/mephalrith/noodles/backend/internal/errs"
	"github.com/mephalrith/noodles/backend/internal/respond"
	"github.com/mephalrith/noodles/backend/internal/services"
)

func HandleListServices(k8s *services.K8sService) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		svcList, err := k8s.DiscoverServices(r.Context())
		if err != nil {
			services.Logger.Error("Service discovery failed", "error", err)
			respond.Error(w, errs.Internal("Failed to discover services"))
			return
		}

		respond.OK(w, svcList)
	}
}
